tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/localedistance/DistanceTable.java

```diff
@@ -0,0 +1,428 @@
+// © 2017 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html#License
+package org.unicode.icu.tool.cldrtoicu.localedistance;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.ibm.icu.impl.locale.LocaleDistance.DISTANCE_SKIP_SCRIPT;
+import static com.ibm.icu.impl.locale.LocaleDistance.IX_DEF_LANG_DISTANCE;
+import static com.ibm.icu.impl.locale.LocaleDistance.IX_DEF_REGION_DISTANCE;
+import static com.ibm.icu.impl.locale.LocaleDistance.IX_DEF_SCRIPT_DISTANCE;
+import static com.ibm.icu.impl.locale.LocaleDistance.IX_LIMIT;
+import static com.ibm.icu.impl.locale.LocaleDistance.IX_MIN_REGION_DISTANCE;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+
+import com.ibm.icu.util.Output;
+
+/**
+ * Table representing the conceptual distance between pairs of languages.
+ *
+ * <p>Note that the builder class is not resuable, and once a table is built, the
+ * builder is invalid. Furthermore, since the table data itself is mutable, care
+ * must be taken to avoid modifying either the Trie or the returned distance array.
+ *
+ * <p>Note that internally the {@code '*'} character used as a wildcard for subtags
+ * is replaced by the {@code '�'} character (a.k.a "ANY"), whenever a subtag is
+ * passed into the API. This is because the underlying Trie structure generated by
+ * the distance table also supports the idea of "star" (which also uses {@code '*'}).
+ * This difference is encapsulated within this class and the {@link Trie} class.
+ */
+// TODO: Finish this once it's clear what it is actually doing.
+final class DistanceTable {
```

Comment:
Equivalent to:
https://github.com/unicode-org/icu/blob/master/icu4j/tools/misc/src/com/ibm/icu/dev/tool/locale/LocaleDistanceBuilder.java#L179

* Split into builder and final value to better encapsulate processing.
* I still don't fully understand what's going on here though.

Reply:
I now understand everything here.